src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -183,13 +183,36 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
 
         private static async Task<Socket> ConnectAsync(string serverName, int port)
         {
-            Socket socket;
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            // On unix we can only connect using IP addresses, rather than host names, using the instance connect methods.
+            IPAddress[] addresses = await Dns.GetHostAddressesAsync(serverName).ConfigureAwait(false);
+            IPAddress targetAddrV4 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetwork));
+            IPAddress targetAddrV6 = Array.Find(addresses, addr => (addr.AddressFamily == AddressFamily.InterNetworkV6));
+            if (targetAddrV4 != null && targetAddrV6 != null)
             {
-                socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                return await ParallelConnectAsync(new IPAddress[] { targetAddrV4, targetAddrV6 }, port).ConfigureAwait(false);
+            }
+            else
+            {
+                Socket socket;
+                IPAddress targetAddr;
```

Comment:
This can be simplified.


You can initialize targetAddr = targetAddrV4 != null ? targetAddrV4 : targetAddrV6;  // Priority to IPv4
This reduce the socket initialization code below to 
socket = Socket(targetAddr.AddressFamily, SocketType.Stream, ProtocolType.Tcp);